pkg/engine/plan/type_field_extractor.go

```diff
@@ -0,0 +1,197 @@
+func (r *TypeFieldExtractor) GetAllNodes() (rootNodes, childNodes []TypeField) {
+	rootNodes = r.getAllRootNodes()
+	childNodes = r.getAllChildNodes(rootNodes)
+	return
+}
+
+func (r *TypeFieldExtractor) getAllRootNodes() []TypeField {
+	var rootNodes []TypeField
+
+	for _, astNode := range r.document.RootNodes {
+		switch astNode.Kind {
+		case ast.NodeKindObjectTypeExtension, ast.NodeKindObjectTypeDefinition:
+			r.addRootNodes(astNode, &rootNodes)
+		}
+	}
+
+	return rootNodes
+}
+
+func (r *TypeFieldExtractor) getAllChildNodes(rootNodes []TypeField) []TypeField {
+	var childNodes []TypeField
+
+	for i := range rootNodes {
+		fieldNameToRef := make(map[string]int, len(rootNodes[i].FieldNames))
+
+		rootNodeASTNode, exists := r.document.Index.FirstNodeByNameStr(rootNodes[i].TypeName)
+		if !exists {
+			continue
+		}
+
+		fieldRefs := r.document.NodeFieldDefinitions(rootNodeASTNode)
+		for _, fieldRef := range fieldRefs {
+			fieldName := r.document.FieldDefinitionNameString(fieldRef)
+			fieldNameToRef[fieldName] = fieldRef
+		}
+
+		for _, fieldName := range rootNodes[i].FieldNames {
+			fieldRef := fieldNameToRef[fieldName]
+
+			fieldTypeName := r.document.NodeNameString(r.document.FieldDefinitionTypeNode(fieldRef))
+			r.findChildNodesForType(fieldTypeName, &childNodes)
+		}
+	}
+
+	return childNodes
+}
+
+func (r *TypeFieldExtractor) findChildNodesForType(typeName string, childNodes *[]TypeField) {
+	node, exists := r.document.Index.FirstNodeByNameStr(typeName)
+	if !exists {
+		return
+	}
+
+	fieldsRefs := r.document.NodeFieldDefinitions(node)
+
+	for _, fieldRef := range fieldsRefs {
+		fieldName := r.document.FieldDefinitionNameString(fieldRef)
+
+		if added := r.addChildTypeFieldName(typeName, fieldName, childNodes); !added {
+			continue
+		}
+
+		fieldTypeName := r.document.NodeNameString(r.document.FieldDefinitionTypeNode(fieldRef))
+		r.findChildNodesForType(fieldTypeName, childNodes)
+	}
+}
+
+func (r *TypeFieldExtractor) addChildTypeFieldName(typeName, fieldName string, childNodes *[]TypeField) bool {
+	for i := range *childNodes {
+		if (*childNodes)[i].TypeName != typeName {
+			continue
+		}
+
+		for _, field := range (*childNodes)[i].FieldNames {
+			if field == fieldName {
+				return false
+			}
+		}
+
+		(*childNodes)[i].FieldNames = append((*childNodes)[i].FieldNames, fieldName)
+		return true
+	}
+
+	*childNodes = append(*childNodes, TypeField{
+		TypeName:   typeName,
+		FieldNames: []string{fieldName},
+	})
+
+	return true
+}
+
+func (r *TypeFieldExtractor) addRootNodes(astNode ast.Node, rootNodes *[]TypeField) {
+	typeName := r.document.NodeNameString(astNode)
+
+	// we need to first build the base schema so that we get a valid Index
+	// to look up if typeName is a RootOperationTypeName
+	// the service SDL itself might use ObjectTypeExtension types which will not be indexed
+	document := r.baseSchema()
+
+	// node should be an entity or a root operation type definition
```

Comment:
maybe move the first line of this comment to a method doc?

smth like
```
// isEntity - checks that node an entity or a root operation type definition and has a federation directive
```

Reply:
ok